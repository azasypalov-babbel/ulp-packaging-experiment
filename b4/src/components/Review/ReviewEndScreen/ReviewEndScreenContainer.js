import { connect } from 'react-redux';
import { compose } from 'redux';
import ReviewEndScreen from './ReviewEndScreen';
import { withServices } from '../../shared/withServices';
import * as accountSelectors from '../../../dux/account/selectors';
import * as sequenceSelectors from '../../../dux/sequence/selectors';
import * as reviewActions from '../../../dux/review/actions';
import * as sequenceActions from '../../../dux/sequence/actions';
import { formatLL } from '../../../lib/markupFormatter';
import playFeedbackSound, { FEEDBACK_SOUND } from '../../../lib/playFeedbackSound';


export const getGrade = (correctItemsCount, totalItemsCount) => {
  const percentage = (correctItemsCount * 100) / totalItemsCount;
  if (percentage > 66) return 'high';
  if (percentage > 33) return 'medium';

  return 'low';
};

const playSound = (soundService, mediaUrlService) => (item) => {
  if (item.sound && item.sound.id) {
    const url = mediaUrlService.soundURL(item.sound.id);
    soundService.play(url);
  }
};

const formatLearnLanguageText = (item) => ({
  ...item,
  learnLanguageText: formatLL(item.learnLanguageText)
});

export const mapStateToProps = ({ account, review, session, sequence }) => ({
  displayName: accountSelectors.displayName(account),
  hasNextReviewItems: review.hasNextReviewItems,
  locale: session.locale,
  purgeableItemsCount: sequenceSelectors.purgeableItems(sequence).length,
  correctItems: sequenceSelectors.correctItems(sequence).map(formatLearnLanguageText),
  incorrectItems: sequenceSelectors.incorrectItems(sequence).map(formatLearnLanguageText)
});

const mapDispatchToProps = {
  onCorrectErrorsButtonClick: sequenceActions.startPurge,
  onReviewMoreButtonClick: reviewActions.fetchNextReviewItems,
  onCloseButtonClick: reviewActions.closeReview
};

export const mergeProps = (stateProps, dispatchProps, ownProps) => {
  const { soundService, mediaUrlService } = ownProps;
  const { correctItems, incorrectItems } = stateProps;

  const totalItemsCount = correctItems.length + incorrectItems.length;
  const correctItemsCount = correctItems.length;
  const grade = getGrade(correctItemsCount, totalItemsCount);

  const playEndScreenSound = () => playFeedbackSound(FEEDBACK_SOUND.END);

  return {
    ...ownProps,
    ...stateProps,
    ...dispatchProps,
    playSound: playSound(soundService, mediaUrlService),
    playEndScreenSound,
    grade,
    totalItemsCount
  };
};

export default compose(
  withServices(['soundService', 'mediaUrlService']),
  connect(mapStateToProps, mapDispatchToProps, mergeProps)
)(ReviewEndScreen);
