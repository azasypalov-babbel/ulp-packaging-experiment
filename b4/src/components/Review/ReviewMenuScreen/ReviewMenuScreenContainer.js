import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import ReviewMenuScreen from './ReviewMenuScreen';
import * as reviewActions from '../../../dux/review/actions';

const mapStateToProps = (state) => ({
  interactions: state.review.interactionTypes.map((type) => type.id)
});

const mapDispatchToProps = {
  fetchInteractionTypes: reviewActions.fetchInteractionTypes,
  updateInteraction: reviewActions.updateInteraction
};

class ReviewMenuScreenContainer extends React.Component {
  constructor(props) {
    super(props);
    this.handleSelectInteraction = this.handleSelectInteraction.bind(this);
  }

  handleSelectInteraction(interaction) {
    this.props.updateInteraction(interaction);
  }

  render() {
    return (
      <ReviewMenuScreen
        onSelectInteraction={this.handleSelectInteraction}
        translations={this.translations}
        interactions={this.props.interactions}
      />
    );
  }
}

ReviewMenuScreenContainer.propTypes = {
  updateInteraction: PropTypes.func.isRequired,
  interactions: PropTypes.arrayOf(PropTypes.string).isRequired
};

export default compose(
  connect(mapStateToProps, mapDispatchToProps)
)(ReviewMenuScreenContainer);
