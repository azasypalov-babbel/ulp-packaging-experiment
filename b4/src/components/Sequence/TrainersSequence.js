import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import TrainerContainer from '../Sequence/TrainerContainer';
import MicPermissionContainer from '../shared/SpeechRecognition/MicPermissionContainer';

import * as sequenceActions from '../../dux/sequence/actions';
import * as trackerActions from '../../dux/tracker/actions';
import * as sessionActions from '../../dux/session/actions';
import { isMicEnabled, isReview } from '../../dux/session/selectors';
import { calculateTrainerProgress, currentTrainerItemIndex } from '../../dux/sequence/selectors';

import * as Engines from '../../services/speechRecognition/engines';
import TrainerErrorBoundary from './TrainerErrorBoundary';

const mapStateToProps = ({ session, sequence, permissions }) => ({
  isMicEnabled: isMicEnabled(session),
  trainers: sequence.trainers,
  isReview: isReview(session),
  trainerProgress: calculateTrainerProgress(sequence.trainers[sequence.currentTrainerIndex]),
  currentTrainerIndex: sequence.currentTrainerIndex,
  isPermissionCompleted: permissions.completed,
  currentTrainerItemIndex: currentTrainerItemIndex(sequence)
});

const mapDispatchToProps = {
  track: trackerActions.track,
  abortSequence: sequenceActions.abortSequence,
  completeTrainer: sequenceActions.completeTrainer,
  allocateItems: sequenceActions.allocateItems,
  skipItem: sequenceActions.skipItem,
  setMicSettings: sessionActions.setMicSettings
};

export class TrainersSequence extends React.Component {
  constructor(props) {
    super(props);

    this.handleTrainerStart = this.handleTrainerStart.bind(this);
    this.handleTrainerFinish = this.handleTrainerFinish.bind(this);
    this.handleTrainerItemSkip = this.handleTrainerItemSkip.bind(this);
    this.unloadHandler = this.unloadHandler.bind(this);
  }

  handleTrainerStart({ scorableItemsCount }) {
    if (scorableItemsCount) {
      this.props.allocateItems(scorableItemsCount);
    }
  }

  handleTrainerItemSkip(item, attempt, trainerData) {
    this.props.skipItem(item, attempt, trainerData);
  }

  handleTrainerFinish() {
    this.props.completeTrainer();
  }

  unloadHandler() {
    this.props.abortSequence();
  }

  componentDidMount() {
    addEventListener('beforeunload', this.unloadHandler);
  }

  componentWillUnmount() {
    removeEventListener('beforeunload', this.unloadHandler);
  }

  renderTrainer() {
    const currentTrainer = this.props.trainers[this.props.currentTrainerIndex];
    const shouldShowToolbar =
      this.props.shouldShowToolbar &&
      [Engines.types.WEB_SPEECH, Engines.types.NATIVE_SPEECH].includes(this.props.speechEngineName) &&
      currentTrainer.interaction === 'speak';

    const key = `${this.props.currentTrainerIndex}_${currentTrainer.interaction}`;

    return (
      <TrainerErrorBoundary
        trainerType={currentTrainer.type}
        trainerInteraction={currentTrainer.interaction}
        onError={this.handleTrainerFinish}
      >
        <TrainerContainer
          key={key}
          isMicEnabled={this.props.isMicEnabled}
          trainer={currentTrainer}
          trainerProgress={this.props.trainerProgress}
          currentTrainerItemIndex={this.props.currentTrainerItemIndex}
          onStart={this.handleTrainerStart}
          onItemSkip={this.handleTrainerItemSkip}
          onFinish={this.handleTrainerFinish}
          speechEngineName={this.props.speechEngineName}
          shouldShowToolbar={shouldShowToolbar}
        />
      </TrainerErrorBoundary>
    );
  }

  render() {
    const currentTrainer = this.props.trainers[this.props.currentTrainerIndex];
    const { isMicEnabled, speechEngineName, isPermissionCompleted } = this.props;

    if (!currentTrainer) return null;

    if (speechEngineName === Engines.types.WEB_SPEECH) {
      const shouldRenderMicPermission =
        isMicEnabled &&
        currentTrainer.interaction === 'speak' &&
        !isPermissionCompleted;

      if (shouldRenderMicPermission) {
        return (
          <MicPermissionContainer speechEngineName={speechEngineName} />
        );
      }
    }

    return (
      this.renderTrainer()
    );
  }
}

TrainersSequence.propTypes = {
  isMicEnabled: PropTypes.bool,
  allocateItems: PropTypes.func.isRequired,
  completeTrainer: PropTypes.func.isRequired,
  trainers: PropTypes.arrayOf(PropTypes.object).isRequired,
  currentTrainerIndex: PropTypes.number,
  speechEngineName: PropTypes.oneOf([
    Engines.types.LEGACY_SPEECH,
    Engines.types.NATIVE_SPEECH,
    Engines.types.WEB_SPEECH
  ]),
  isPermissionCompleted: PropTypes.bool.isRequired,
  trainerProgress: PropTypes.number,
  shouldShowToolbar: PropTypes.bool.isRequired,
  skipItem: PropTypes.func.isRequired,
  currentTrainerItemIndex: PropTypes.number,
  abortSequence: PropTypes.func.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(TrainersSequence);
