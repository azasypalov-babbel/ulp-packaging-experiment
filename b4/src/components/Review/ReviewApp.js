import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import * as reviewActions from '../../dux/review/actions';
import * as statisticsActions from '../../dux/statistics/actions';
import NavbarContainer from '../shared/Navbar/NavbarContainer';
import ApplicationLayout from '../shared/ApplicationLayout';
import ReviewMenuScreenContainer from './ReviewMenuScreen/ReviewMenuScreenContainer';
import ReviewEndScreenContainer from './ReviewEndScreen/ReviewEndScreenContainer';
import TrainersSequence from '../Sequence/TrainersSequence';
import FeatureToggles from '../shared/FeatureToggles';
import Loader from '../shared/Loader';
import CenterContent from '../shared/CenterContent';
import * as accountActions from '../../dux/account/actions';
import * as trackerActions from '../../dux/tracker/actions';
import * as permissionsActions from '../../dux/permissions/actions';
import MessagesContainer from '../shared/Messages/MessagesContainer';
import { withServices } from '../shared/withServices';
import withFontsLoaded from '../shared/withFontsLoaded';
import { BottomLayout } from '../shared/BottomLayout';

import ApplicationErrorBoundary from '../ApplicationErrorBoundary';

import { performancePageLoadedEvent } from '../../dux/tracker/events/performance';
import TrainerOverlay from '../shared/TrainerOverlay';

const mapStateToProps = (state) => ({
  state: state,
  isLoading: Boolean(state.review.loading || state.statistics.loading),
  started: state.sequence.started,
  completed: state.sequence.completed,
  selectedInteraction: state.review.selectedInteraction,
  locale: state.session.locale,
  learnLanguageAlpha3: state.session.learnLanguageAlpha3
});

export const mapDispatchToProps = {
  fetchAccount: accountActions.fetchAccount,
  fetchTrainerItemsStatistics: statisticsActions.fetchTrainerItemsStatistics,
  fetchInteractionTypes: reviewActions.fetchInteractionTypes,
  track: trackerActions.track,
  initMicPermissions: permissionsActions.initMicPermissions
};

export class ReviewApp extends React.Component {
  componentDidMount() {
    this.initReview();
  }

  async initReview() {
    const {
      fetchAccount,
      fetchInteractionTypes,
      fetchTrainerItemsStatistics,
      initMicPermissions,
      track,
      locale,
      learnLanguageAlpha3
    } = this.props;

    await fetchAccount();

    /* eslint-disable camelcase */
    track({
      event: 'product:view_shown',
      version: 1,
      payload: {
        view_name: 'review',
        locale: locale,
        learn_language_alpha3: learnLanguageAlpha3
      }
    });
    /* eslint-enable camelcase */

    track(performancePageLoadedEvent({ locale, learnLanguageAlpha3 }));

    fetchInteractionTypes();
    fetchTrainerItemsStatistics();

    initMicPermissions();
  }

  renderSequence() {
    const {
      started,
      completed,
      speechRecognitionService,
      isLoading,
      isLoadingFonts,
      selectedInteraction
    } = this.props;

    if (isLoading || isLoadingFonts) {
      return <CenterContent><Loader /></CenterContent>;
    }

    if (!selectedInteraction) {
      return <ReviewMenuScreenContainer />;
    }

    if (started && !completed) {
      return <TrainersSequence
        speechEngineName={speechRecognitionService?.getEngineName()}
        shouldShowToolbar={false}
      />;
    }

    if (completed) {
      return <ReviewEndScreenContainer />;
    }

    return null;
  }

  render() {
    return (
      <ApplicationLayout>
        <ApplicationErrorBoundary name="ReviewApp">
          <NavbarContainer />
          <FeatureToggles />
          <MessagesContainer />
          <TrainerOverlay />
          <main>
            {this.renderSequence()}
          </main>
          <BottomLayout />
        </ApplicationErrorBoundary>
      </ApplicationLayout>
    );
  }
}

ReviewApp.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  isLoadingFonts: PropTypes.bool.isRequired,
  fetchAccount: PropTypes.func.isRequired,
  fetchTrainerItemsStatistics: PropTypes.func.isRequired,
  initMicPermissions: PropTypes.func.isRequired,
  selectedInteraction: PropTypes.string,
  fetchInteractionTypes: PropTypes.func.isRequired,
  track: PropTypes.func.isRequired,
  locale: PropTypes.string.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired,
  speechRecognitionService: PropTypes.any.isRequired,
  started: PropTypes.bool,
  completed: PropTypes.bool
};

export default compose(
  withServices(['speechRecognitionService']),
  withFontsLoaded,
  connect(mapStateToProps, mapDispatchToProps)
)(ReviewApp);
