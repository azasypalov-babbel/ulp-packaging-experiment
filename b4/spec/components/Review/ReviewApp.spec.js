import React from 'react';
import { shallow } from 'enzyme';
import ReviewMenuScreenContainer from '../../../src/components/Review/ReviewMenuScreen/ReviewMenuScreenContainer';
import ReviewEndScreenContainer from '../../../src/components/Review/ReviewEndScreen/ReviewEndScreenContainer';
import TrainersSequence from '../../../src/components/Sequence/TrainersSequence';
import { ReviewApp, mapDispatchToProps } from '../../../src/components/Review/ReviewApp';
import mockSpeechRecognitionService from '../../../src/services/speechRecognition/service';

jest.mock('../../../src/dux/tracker/events/vocabularyReview', () => ({
  vocabReviewEndedAbortEvent: () => ({ event: 'vocabulary_review:ended:abort' })
}));

jest.mock('../../../src/providers/b3', () => (Component) => (props) => <Component {...props} />);

describe('<ReviewApp />', () => {
  const defaultProps = {
    completed: false,
    fetchAccount: jest.fn(() => Promise.resolve()),
    fetchInteractionTypes: jest.fn(() => Promise.resolve()),
    fetchTrainerItemsStatistics: jest.fn(() => Promise.resolve()),
    initMicPermissions: jest.fn(() => Promise.resolve()),
    isLoading: false,
    isLoadingFonts: false,
    learnLanguageAlpha3: 'FRA',
    locale: 'en',
    speechRecognitionService: mockSpeechRecognitionService,
    started: true,
    track: jest.fn().mockResolvedValue({})
  };

  test('#mapDispatchToProps', () => {
    expect(mapDispatchToProps).toMatchSnapshot();
  });

  describe('loading state', () => {
    describe('when loading', () => {
      let wrapper;

      beforeEach(() => {
        wrapper = shallow(<ReviewApp {...defaultProps} isLoading={true} />);
      });

      it('renders loader', () => {
        expect(wrapper.find('Loader')).toHaveLength(1);
      });
    });

    describe('when is not loading and fonts loaded', () => {
      let wrapper;

      beforeEach(() => {
        wrapper = shallow(<ReviewApp {...defaultProps} isLoading={false} />);
      });

      it('does not render loader', () => {
        expect(wrapper.find('Loader')).toHaveLength(0);
      });
    });

    describe('when is not loading but fonts are loading', () => {
      let wrapper;

      beforeEach(() => {
        wrapper = shallow(<ReviewApp {...defaultProps} isLoading={false} isLoadingFonts={true} />);
      });

      it('does not render loader', () => {
        expect(wrapper.find('Loader')).toHaveLength(1);
      });
    });
  });

  describe('component did mount', () => {
    beforeEach(() => {
      shallow(<ReviewApp {...defaultProps} />);
    });

    test('fetches interaction types', () => {
      expect(defaultProps.fetchInteractionTypes).toHaveBeenCalled();
    });

    test('fetches trainer items statistics', () => {
      expect(defaultProps.fetchTrainerItemsStatistics).toHaveBeenCalled();
    });

    test('fetches account information', () => {
      expect(defaultProps.fetchAccount).toHaveBeenCalled();
    });

    test('initialises mic permissions', () => {
      expect(defaultProps.initMicPermissions).toHaveBeenCalled();
    });

    test('should track product:view_shown', () => {
      expect(defaultProps.track).toHaveBeenCalledWith({
        event: 'product:view_shown',
        version: 1,
        payload: {
          view_name: 'review',
          locale: 'en',
          learn_language_alpha3: 'FRA'
        }
      });
    });
  });

  describe('when interaction is not selected', () => {
    test('renders ReviewMenu', () => {
      const component = shallow(<ReviewApp
        {...defaultProps}
        selectedInteraction={null}
      />);

      expect(component.find(ReviewMenuScreenContainer)).toHaveLength(1);
    });
  });

  describe('when interaction is selected', () => {
    test('renders TrainersSequence', () => {
      const component = shallow(<ReviewApp
        {...defaultProps}
        selectedInteraction="write"
      />);

      expect(component.find(TrainersSequence)).toHaveLength(1);
    });

    describe('when the sequence is completed', () => {
      test('renders ReviewEndScreenContainer', () => {
        const component = shallow(<ReviewApp
          {...defaultProps}
          completed={true}
          selectedInteraction="write"
        />);

        expect(component.find(ReviewEndScreenContainer)).toHaveLength(1);
      });
    });
  });
});
