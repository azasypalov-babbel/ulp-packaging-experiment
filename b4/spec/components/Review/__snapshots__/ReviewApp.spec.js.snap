// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReviewApp /> #mapDispatchToProps 1`] = `
Object {
  "fetchAccount": [Function],
  "fetchInteractionTypes": [Function],
  "fetchTrainerItemsStatistics": [Function],
  "initMicPermissions": [Function],
  "track": [Function],
}
`;
